upstairs/src/volume.rs

```diff
@@ -1096,25 +1101,35 @@ impl Volume {
     }
 
     // We compare two VolumeConstructionRequests to see if the second one
-    // can be a replacement for the first.
-    // The requirements are:
+    // could be considered as an updated VCR to the original.  We are
+    // considering two situations here, one for migration and one for a
+    // downstairs target replacement.
+    //
+    // The requirements to allow a new VCR are:
     // 1. Only VolumeConstructionRequests::Volume type is supported.
     // 2. Sub volumes must all be VolumeConstructionRequest::Region
     // 3. Everything must be the same between the two Volumes, except:
     //    A. The new generation number must be greater than the old.
-    //    B. Only one CrucibleOpts::target[] is different.
-    //    C. The new Volume can have None for read only parent if the
+    //    B. The new Volume can have None for read only parent if the
     //       original had Some(), or it must match the Some().
     //
+    //  Specifically for the downstairs target replacement case we also require
+    //  this difference:
+    //    C. Only one CrucibleOpts::target[] is different.
+    //
     // Any other difference between the two volumes is considered a failure.
     //
-    // On success, we record the old and new downstairs target and return
-    // those to the caller.
-    pub async fn compare_vcr_for_replacement(
+    // We return Ok(None) if requirements 1, 2, 3A, 3B are met.  This would
```

Comment:
This makes sense to me, since we only need to return a socket address pair if we're replacing, but it feels a little weird to me to have the caller check for the flavor of `Ok` it got back when it could have specified to the callee which checks to do. That is, I think the downstairs-replacement path is going to end up with something like this:

```rust
match compare_vcr_for_update(original, replacement, log)? {
    Ok(Some(_)) => todo!("okay to proceed"),
    Ok(None) => bail!("target replacement check failed"),
}
```

Might it make sense to have two functions? `compare_vcr_for_update` checks 1, 2, 3A, and 3B and returns a `Result<()>`; migration calls that one. `compare_vcr_for_replacement` calls `compare_vcr_for_update` and, if that succeeds, does check 3C, and if that succeeds it returns `Result<(SocketAddr, SocketAddr)>`. Then the downstairs replacement code can just `let (old_addr, new_addr) = compare_vcr_for_replacement()?;`.

WDYT?

Reply:
The challenge here is that, and perhaps I need to make it clear in the comment as well, is:
* For migration, I don't want 3C, the targets must be the same.  
* For replacement, I do want 3C, the targets must be different.  

The `compare_vcr_for_replacement` does not build on top of `compare_vcr_for_migration`

I think I can do it with a wrapper in front of `compare_vcr_for_update`, where I just hide the `Ok` check like you have shown above and have a `migration_check` and a `replacement_check`, 
`migration_check` just looks for `Ok(None)`, and `replacement_check` wants the `Ok(Some(_))`
Then I'm at least sweeping the `Ok` check under the rug and the callers don't have to match.

I could also make a "check everything except the `sub_volume` targets" function (what is common between the replace/migrate cases). If I did that I would have to make another two functions that then did the translation/selection from `VolumeConstructionRequest` -> `VcrVolume` -> `SubVolume` -> Target Vec, and returned the Ok/Err depending on what they wanted, which would then look something like this:

```
fn compare_vcr_for_migration(..) -> Result<()> {
   compare_vcr_common(..)?
   compare_vcr_subtarget_the_same()
}
```
and
```
fn compare_vcr_for_migration(..) -> Result<(SocketAddr, SocketAddr)> {
   compare_vcr_common(..)?
   compare_vcr_subtarget_one_different()
}
```

Or, maybe I make a `give_me_the_subvolume_target_from_this_vcr` function, then compare the two targets after I have done the `compare_vcr_common`, but.. I don't know what ends up being the better solution here.   I kind of like just hiding the Ok check. 

Let me know if you feel strongly about either of these two solutions (or some third idea).  I do think it's worth updating this to be a little better and having two functions to call depending on what the caller wants, but I need a little direction on which path you think might be better here.

Reply:
I went ahead and tried option A (the wrapper around `compare_vcr_for_update`)
Let me know what you think.